### PR TITLE
[SPARK-37675][SPARK-37793] Prevent overwriting of push shuffle merged files once the shuffle is finalized

### PR DESCRIPTION
**Pull request by otterc**
### What changes were proposed in this pull request?
This fixes the bugs that were reported in SPARK-37675 and SPARK-37793.
- Empty merged partitions were reported by the shuffle server to the driver.
- The push merged files were getting overwritten after a shuffle merge is finalized. 
- Throwing exception in the finalization of a shuffle for which the shuffle server didn't receive any blocks.

### Why are the changes needed?
Changes are need to fix the bug.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Have added unit test.

### CODE REVIEW

**Comment by pan3793** [comment]
I tested your first commit 388559f, it does not solve the problem.
image

**Comment by otterc** [comment]
@pan3793 Thanks for testing this. So with this patch are you still seeing that the executor is requesting for merged blocks which either had no data or were overwritten by the shuffle service? 
Will you be able to share the logs with me again? It will be great if you can turn the debug level logging on for the `RemoteBlockPushResolver`.

**Comment by pan3793** [comment]
@otterc Oops, sorry, I forgot to restart the node manager so the patch does not take effect, let me retry again.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java [hunk 1]
```diff
@@ -1008,19 +995,28 @@ AppShufflePartitionInfo getPartitionInfo() {
...
   public static class AppShuffleMergePartitionsInfo {
+    // ConcurrentHashMap doesn't allow null for keys or values which is why this is required.
+    // Marker to identify finalized indeterminate shuffle partitions in the case of indeterminate
+    // stage retries.
+    @VisibleForTesting
+    public static final Map<Integer, AppShufflePartitionInfo> SHUFFLE_FINALIZED_MARKER =
+        Collections.emptyMap();
     private final int shuffleMergeId;
-    private final Map<Integer, AppShufflePartitionInfo> shuffleMergePartitions;
+    private final Map<Integer, AppShufflePartitionInfo> mergePartitions;
```
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1005]: nit: avoid rename here ?

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java [hunk 2]
```diff
@@ -1287,6 +1287,52 @@ void closeAndDeletePartitionFiles(Map<Integer, AppShufflePartitionInfo> partitio
       + " up", appShuffleInfo.getMergedShuffleDataFile(0, 4, 0).exists());
   }
 
+  // Test for SPARK-37675 and SPARK-37793
+  @Test
+  public void testAllBlocksAreRejectedWhenReceivedAfterFinalization() throws IOException {
```
Comment (mridulm) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:1292]: Also add a test for the cardinality change ?

**Review by mridulm — COMMENTED** [review]
Looks good, just couple of comments.

**Comment by mridulm** [comment]
@pan3793 Functionally, @otterc's patch looks good (except for testing enhancements, renames, etc).
Can you also confirm if this works for you ? Thanks !

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java (hunk 2 above)
Comment (otterc) [reply common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:1306]: Added the test.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java (hunk 1 above)
Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: reverted the rename

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java [hunk 3]
```diff
@@ -534,35 +515,33 @@ public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) {
...
-    if (msg.shuffleMergeId == DETERMINATE_SHUFFLE_MERGE_ID) {
-      AppShuffleMergePartitionsInfo appShuffleMergePartitionsInfo =
-        appShuffleInfo.shuffles.remove(msg.shuffleId);
-      if (appShuffleMergePartitionsInfo != null) {
-        shuffleMergePartitionsRef.set(appShuffleMergePartitionsInfo.shuffleMergePartitions);
+    appShuffleInfo.shuffles.compute(msg.shuffleId, (shuffleId, mergePartitionsInfo) -> {
+      if (null == mergePartitionsInfo) {
+        //If the mergePartitions was never created then it means that there weren't any push
+        //blocks that were ever received for this shuffle. This could be the case when the driver
+        //doesn't wait for enough time to start the stage which reads this shuffle data.
```
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:522]: nit: space between `//` and text.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java [hunk 4]
```diff
@@ -534,35 +515,33 @@ public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) {
...
+        return new AppShuffleMergePartitionsInfo(msg.shuffleMergeId, true);
+      } else if (msg.shuffleMergeId < mergePartitionsInfo.shuffleMergeId
+          || mergePartitionsInfo.isFinalized()) {
```
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:525]: nit: move `||` to prev line

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java [hunk 5]
```diff
@@ -534,35 +515,33 @@ public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) {
...
+        mergedShuffleCleaner.execute(() ->
+            closeAndDeletePartitionFiles(mergePartitionsInfo.shuffleMergePartitions));
+        return new AppShuffleMergePartitionsInfo(msg.shuffleMergeId, true);
+      } else {
+        // This block covers:
+        //  1. finalization of determinate stage
+        //  2. finalization of indeterminate stage if the shuffleMergeId related to it is the one
+        //  for which the message is received.
+        shuffleMergePartitionsRef.set(mergePartitionsInfo.shuffleMergePartitions);
+        return new AppShuffleMergePartitionsInfo(msg.shuffleMergeId, true);
```
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:542]: nit: All paths in this lambda, which are not throwing exceptions, are doing the same thing - `new AppShuffleMergePartitionsInfo(msg.shuffleMergeId, true);`
Make that common ?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java [hunk 6]
```diff
@@ -576,14 +555,25 @@ public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) {
...
             partition.finalizePartition();
-            bitmaps.add(partition.mapTracker);
-            reduceIds.add(partition.reduceId);
-            sizes.add(partition.getLastChunkOffset());
+            if (partition.mapTracker.getCardinality() > 0) {
+              bitmaps.add(partition.mapTracker);
+              reduceIds.add(partition.reduceId);
+              sizes.add(partition.getLastChunkOffset());
+              logger.debug("{} attempt {} shuffle {} shuffleMerge {}: finalization results added"
+              + " for partition {} data size {} index size {} meta size {}",
```
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:568]: nit: move the `+` to prev line (here and elsewhere in string concat for log msgs)

**Review by mridulm — APPROVED** [review]
Just some minor style comments, looks good to me.
+CC @Ngone51

**Comment by mridulm** [comment]
@otterc Can you remove the WIP tag, given the PR should be complete now ? Thx

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java [hunk 7]
```diff
@@ -1008,20 +996,28 @@ AppShufflePartitionInfo getPartitionInfo() {
    * required for the shuffles of indeterminate stages.
    */
   public static class AppShuffleMergePartitionsInfo {
+    // ConcurrentHashMap doesn't allow null for keys or values which is why this is required.
+    // Marker to identify finalized indeterminate shuffle partitions in the case of indeterminate
```
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1000]: Nit: remove indeterminate

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java [hunk 8]
```diff
@@ -1161,8 +1161,8 @@ public void testFinalizeOfDeterminateShuffle() throws IOException {
 
     RemoteBlockPushResolver.AppShuffleInfo appShuffleInfo =
       pushResolver.validateAndGetAppShuffleInfo(TEST_APP);
-    assertTrue("Metadata of determinate shuffle should be removed after finalize shuffle"
-      + " merge", appShuffleInfo.getShuffles().get(0) == null);
+    assertTrue("Determinate shuffle should be marked finalized"
+      + " merge", appShuffleInfo.getShuffles().get(0).isFinalized());
```
Comment (otterc) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:1165]: Nit: remove ` merge`

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java [hunk 9]
```diff
@@ -1287,6 +1287,79 @@ void closeAndDeletePartitionFiles(Map<Integer, AppShufflePartitionInfo> partitio
...
 
+  @Test
+  public void testFinalizationResultIsEmptyWhenTheServerDidNotReceiveAnyBlocks() {
+    //shuffle 1 0 is finalized even though the server didn't receive any blocks for it.
+    MergeStatuses statuses = pushResolver.finalizeShuffleMerge(
+        new FinalizeShuffleMerge(TEST_APP, NO_ATTEMPT_ID, 1, 0));
+    assertEquals("no partitions were merged", 0, statuses.reduceIds.length);
+    RemoteBlockPushResolver.AppShuffleInfo appShuffleInfo =
+        pushResolver.validateAndGetAppShuffleInfo(TEST_APP);
+    assertTrue("Determinate shuffle should be marked finalized merge",
```
Comment (otterc) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java:1298]: Nit: fix the msg

**Review by mridulm — APPROVED** [review]
Looks good to me.
Will wait for @pan3793 to confirm it works before merging (and tests success ofcourse).

**Comment by pan3793** [comment]
Thanks @otterc and @mridulm, I verified this patch with 2 rounds of 1TB TPC-DS query, the issue is gone.

One minor question, can we suppress this warning message? I think it's a little bit noisy for users.

```
01-27 16:18:35 WARN [shuffle-client-8-5] scheduler.DAGScheduler@94: Exception encountered when trying to finalize shuffle merge on beta-spark2 for shuffle 189
java.lang.RuntimeException: java.lang.RuntimeException: Shuffle merge finalize request for shuffle 189 with shuffleMergeId 0 is stale shuffle finalize request as shuffle blocks of a higher shuffleMergeId for the shuffle is already being pushed
	at org.apache.spark.network.shuffle.RemoteBlockPushResolver.lambda$finalizeShuffleMerge$11(RemoteBlockPushResolver.java:523)
	at java.util.concurrent.ConcurrentHashMap.compute(ConcurrentHashMap.java:1853)
	at org.apache.spark.network.shuffle.RemoteBlockPushResolver.finalizeShuffleMerge(RemoteBlockPushResolver.java:519)
	at org.apache.spark.network.shuffle.ExternalBlockHandler.handleMessage(ExternalBlockHandler.java:218)
	at org.apache.spark.network.shuffle.ExternalBlockHandler.receive(ExternalBlockHandler.java:122)
	at org.apache.spark.network.server.TransportRequestHandler.processRpcRequest(TransportRequestHandler.java:163)
	at org.apache.spark.network.server.TransportRequestHandler.handle(TransportRequestHandler.java:109)
	at org.apache.spark.network.server.TransportChannelHandler.channelRead0(TransportChannelHandler.java:140)
	at org.apache.spark.network.server.TransportChannelHandler.channelRead0(TransportChannelHandler.java:53)
	at org.sparkproject.io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at org.sparkproject.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at org.sparkproject.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at org.sparkproject.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at org.sparkproject.io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at org.sparkproject.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at org.sparkproject.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at org.sparkproject.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at org.sparkproject.io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at org.sparkproject.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at org.sparkproject.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at org.sparkproject.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at org.apache.spark.network.util.TransportFrameDecoder.channelRead(TransportFrameDecoder.java:102)
	at org.sparkproject.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at org.sparkproject.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at org.sparkproject.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at org.sparkproject.io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at org.sparkproject.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at org.sparkproject.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at org.sparkproject.io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at org.sparkproject.io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at org.sparkproject.io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722)
	at org.sparkproject.io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658)
	at org.sparkproject.io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584)
	at org.sparkproject.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
	at org.sparkproject.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at org.sparkproject.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at org.sparkproject.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)

	at org.apache.spark.network.client.TransportResponseHandler.handle(TransportResponseHandler.java:209) [spark-network-common_2.12-3.3.0-SNAPSHOT.jar:3.3.0-SNAPSHOT]
	at org.apache.spark.network.server.TransportChannelHandler.channelRead0(TransportChannelHandler.java:142) [spark-network-common_2.12-3.3.0-SNAPSHOT.jar:3.3.0-SNAPSHOT]
	at org.apache.spark.network.server.TransportChannelHandler.channelRead0(TransportChannelHandler.java:53) [spark-network-common_2.12-3.3.0-SNAPSHOT.jar:3.3.0-SNAPSHOT]
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286) [netty-handler-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) [netty-codec-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at org.apache.spark.network.util.TransportFrameDecoder.channelRead(TransportFrameDecoder.java:102) [spark-network-common_2.12-3.3.0-SNAPSHOT.jar:3.3.0-SNAPSHOT]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496) [netty-transport-4.1.72.Final.jar:4.1.72.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986) [netty-common-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.73.Final.jar:4.1.73.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_312]
```

**Comment by pan3793** [comment]
Thanks @otterc, would you please add some description of how the corner case caused the file to get overwritten to help others understand how the patch works?

**Comment by mridulm** [comment]
Thanks for verifying @pan3793  !
I will leave the PR open until tomorrow in case there are other comments/reviews, before merging.

**Comment by mridulm** [comment]
@otterc, can you please update the jira with details of root cause ? That should help @pan3793 and future investigations understand the issue fixed here better. Thanks !

**Comment by otterc** [comment]
- The test failures seem to be unrelated to this PR
- @mridulm @pan3793 I have updated the description of SPARK-37675

**Comment by mridulm** [comment]
@otterc Can you take a look at the test failure ? Thx

**Comment by mridulm** [comment]
Merged to master, branch-3.2
Thanks for working on this @otterc 
Thanks for reporting and helping test this @pan3793 !

**Pull request closed without merging**